numpy/core/src/umath/logical_gufuncs.c.src

```diff
@@ -0,0 +1,513 @@
+            @TYPE@ b = (*(@TYPE@ *)b_i) @ISBOOL@;
+            if (@INV@@OP_PREFIX@@OP@(a, b)) {
+                a_i += a_I;
+                b_i += b_I;
+            } else {
+                *((npy_bool *)c_n) = !@ALL_OR_ANY@;
+                break;
+            }
+        }
+
+        a_n += a_N;
+        b_n += b_N;
+        c_n += c_N;
+    }
+}
+#endif
+
+#undef DO_VECTORIZE
+
+/**end repeat3**/
+/**end repeat2**/
+/**end repeat1**/
+/**end repeat**/
+
+/**begin repeat
+ * #TYPE = npy_datetime, npy_timedelta#
+ */
+
+/**begin repeat1
+ * #NAME = all,any#
+ * #ALL_OR_ANY = NPY_TRUE,NPY_FALSE#
+ * #INV = ,!#
+ */
+
+/**begin repeat2
+ * #OPNAME = equal,less,less_equal,greater,greater_equal#
+ * #OP = ==,<,<=,>,>=#
+ */
+
+static  void
+@TYPE@_@NAME@_@OPNAME@(char **args, npy_intp *dimensions,
+                       npy_intp* steps, void* data)
+{
+    npy_intp n;
+    npy_intp N = dimensions[0], nI = dimensions[1];
+    char *a_n = args[0], *b_n = args[1], *c_n = args[2];
+    npy_intp a_N = steps[0], b_N = steps[1], c_N = steps[2];
+    npy_intp a_I = steps[3], b_I = steps[4];
+
+    for (n = 0; n < N; n++) {
+        npy_intp i;
+        char * a_i = a_n;
+        char * b_i = b_n;
+
+        *((npy_bool *)c_n) = @ALL_OR_ANY@;
+
+        for (i=0; i < nI; i++) {
+            @TYPE@ a = *(@TYPE@ *)a_i;
+            @TYPE@ b = *(@TYPE@ *)b_i;
+            npy_bool res = @INV@(a @OP@ b);
+
+            if ((a == NPY_DATETIME_NAT || b == NPY_DATETIME_NAT) && @INV@res) {
+                NPY_ALLOW_C_API_DEF;
+                NPY_ALLOW_C_API;
+                /* 2016-01-18, 1.11 */
+                if (DEPRECATE_FUTUREWARNING(
+                                "In the future, 'NAT @OP@ x' and 'x @OP@ NAT' "
+                                "will always be False.") < 0) {
+                    NPY_DISABLE_C_API;
+                    return;
+                }
+                NPY_DISABLE_C_API;
+            }
+
+            if (res) {
+                a_i += a_I;
+                b_i += b_I;
+            } else {
+                *((npy_bool *)c_n) = !@ALL_OR_ANY@;
+                break;
+            }
+        }
+
+        a_n += a_N;
+        b_n += b_N;
+        c_n += c_N;
+    }
+}
+
+/**end repeat2**/
+
+static  void
+@TYPE@_@NAME@_not_equal(char **args, npy_intp *dimensions,
+                        npy_intp* steps, void* data)
+{
+    npy_intp n;
+    npy_intp N = dimensions[0], nI = dimensions[1];
+    char *a_n = args[0], *b_n = args[1], *c_n = args[2];
+    npy_intp a_N = steps[0], b_N = steps[1], c_N = steps[2];
+    npy_intp a_I = steps[3], b_I = steps[4];
+
+    for (n = 0; n < N; n++) {
+        npy_intp i;
+        char * a_i = a_n;
+        char * b_i = b_n;
+
+        *((npy_bool *)c_n) = @ALL_OR_ANY@;
+
+        for (i=0; i < nI; i++) {
+            @TYPE@ a = *(@TYPE@ *)a_i;
+            @TYPE@ b = *(@TYPE@ *)b_i;
+            npy_bool res = @INV@(a != b);
+
+            if (a == NPY_DATETIME_NAT && a == NPY_DATETIME_NAT) {
```

Comment:
yes nice catch, the same mistake that was done in the original code ... the tests should have caught that